relay_timer_pool.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package tchannel
+
+import (
+	"math"
+	"sync"
+	"time"
+)
+
+type relayTimerTrigger func(items *relayItems, id uint32, isOriginator bool)
+
+type relayTimerPool struct {
+	pool    sync.Pool
+	trigger relayTimerTrigger
+}
+
+type relayTimer struct {
+	pool  *relayTimerPool
+	timer *time.Timer
+
+	// Per-timer paramters passed back when the timer is triggered.
+	items        *relayItems
+	id           uint32
+	isOriginator bool
+}
+
+func (rt *relayTimer) OnTimer() {
+	rt.pool.trigger(rt.items, rt.id, rt.isOriginator)
+}
+
+func newRelayTimerPool(trigger relayTimerTrigger) *relayTimerPool {
+	return &relayTimerPool{
+		trigger: trigger,
+	}
+}
+
+// Get returns a Timer that has not started. Timers must be started explicitly
+// using the Start function.
+func (tp *relayTimerPool) Get() *relayTimer {
+	timer, ok := tp.pool.Get().(*relayTimer)
+	if ok {
+		return timer
+	}
+
+	rt := &relayTimer{
+		pool: tp,
+	}
+	rt.timer = time.AfterFunc(time.Duration(math.MaxInt64), rt.OnTimer)
+	if !rt.timer.Stop() {
+		panic("failed to stop timer set for 1 hour")
```

Comment:
A comment would be nice here, just to explain that we're creating a timer to avoid nil checks later on. On first seeing this, I spent a while trying to figure out if there was a more complex reason to have this logic here.

Assuming that I'm right, and that this is intended to avoid nil checks elsewhere, is it worth the cost? I don't really understand the implementations of the `startTimer` and `stopTimer` functions in the stdlib's `time/sleep.go`, but I assume that they're inserting and removing the timer from the heap, which I assume requires locking and unlocking it. In the context of this relay, that seems a little expensive. (Yes, that's a lot of assumptions.)

Lastly, is this message accurate? It looks like we've set the timer for *much* longer than an hour.